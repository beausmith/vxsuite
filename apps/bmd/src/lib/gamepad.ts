import mod from '../utils/mod'

function getFocusableElements(): HTMLElement[] {
  const tabbableElements = Array.from(
    document.querySelectorAll('button:not([tabindex="-1"]), input')
  )
  const ariaHiddenTabbableElements = Array.from(
    document.querySelectorAll(
      '[aria-hidden="true"] button:not([tabindex="-1"]), [aria-hidden="true"] input'
    )
  )
  return tabbableElements.filter(
    element => ariaHiddenTabbableElements.indexOf(element) === -1
  ) as HTMLElement[]
}

const getActiveElement = () => document.activeElement! as HTMLInputElement

function handleArrowUp() {
  const focusable = getFocusableElements()
  const currentIndex = focusable.indexOf(getActiveElement())
  if (currentIndex > -1) {
    focusable[mod(currentIndex - 1, focusable.length)].focus()
  } else {
    focusable[focusable.length - 1].focus()
  }
}

function handleArrowDown() {
  const focusable = getFocusableElements()
  const currentIndex = focusable.indexOf(getActiveElement())
  focusable[mod(currentIndex + 1, focusable.length)].focus()
}

function handleArrowLeft() {
  const prevButton = document.getElementById('previous') as HTMLButtonElement
  if (prevButton) {
    prevButton.click()
  }
}

function handleArrowRight() {
  const nextButton = document.getElementById('next') as HTMLButtonElement
  if (nextButton) {
    nextButton.click()
  }
}

function handleClick() {
  getActiveElement().click()
}

export function handleGamepadButtonDown(buttonName: string) {
  switch (buttonName) {
    case 'DPadUp':
      handleArrowUp()
      break
    case 'B':
    case 'DPadDown':
      handleArrowDown()
      break
    case 'DPadLeft':
      handleArrowLeft()
      break
    case 'DPadRight':
      handleArrowRight()
      break
    case 'A':
      handleClick()
      break
    // no default
  }
}

// Add Cypress tests if this solution will become permanent
// https://docs.cypress.io/api/commands/type.html
export /* istanbul ignore next */ function handleGamepadKeyboardEvent(
  event: KeyboardEvent
) {
  switch (event.key) {
    case 'ArrowUp':
      handleArrowUp()
      break
    case '[':
    case 'ArrowDown':
      handleArrowDown()
      break
    case 'ArrowLeft':
      handleArrowLeft()
      break
    case 'ArrowRight':
      handleArrowRight()
      break
    case ']':
      handleClick()
      break
    // no default
  }
}
