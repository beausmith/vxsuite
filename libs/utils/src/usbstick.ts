import { strict as assert } from 'assert';
import { sleep } from './sleep';

export const FLUSH_IO_DELAY_MS = 10_000;

function isAvailable() {
  return !!window.kiosk;
}

export enum UsbDriveStatus {
  notavailable = 'notavailable',
  absent = 'absent',
  present = 'present',
  mounted = 'mounted',
  recentlyEjected = 'recentlyEjected',
  ejecting = 'ejecting',
}

async function getDevice(): Promise<KioskBrowser.UsbDrive | undefined> {
  return (await window.kiosk?.getUsbDrives())?.[0];
}

export async function getDevicePath(): Promise<string | undefined> {
  const device = await getDevice();
  return device?.mountPoint;
}

export async function getStatus(): Promise<UsbDriveStatus> {
  if (!isAvailable()) {
    return UsbDriveStatus.notavailable;
  }

  const device = await getDevice();

  if (!device) {
    return UsbDriveStatus.absent;
  }

  if (device.mountPoint) {
    return UsbDriveStatus.mounted;
  }
  return UsbDriveStatus.present;
}

export async function doMount(): Promise<void> {
  const device = await getDevice();
  if (!device || device.mountPoint) {
    return;
  }

  assert(window.kiosk);
  await window.kiosk.mountUsbDrive(device.deviceName);
}

export async function doUnmount(): Promise<void> {
  const device = await getDevice();
  if (!device?.mountPoint) {
    return;
  }
  assert(window.kiosk);
  await window.kiosk.unmountUsbDrive(device.deviceName);
  return await sleep(FLUSH_IO_DELAY_MS);
}
