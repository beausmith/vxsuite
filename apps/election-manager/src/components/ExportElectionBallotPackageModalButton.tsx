import React, { useCallback, useContext, useEffect, useState } from 'react'
import pluralize from 'pluralize'
import styled from 'styled-components'
import { getElectionLocales } from '@votingworks/ballot-encoder'

import { DEFAULT_LOCALE } from '../config/globals'
import {
  getBallotPath,
  getPrecinctById,
  getHumanBallotLanguageFormat,
} from '../utils/election'

import AppContext from '../contexts/AppContext'
import HandMarkedPaperBallot from './HandMarkedPaperBallot'
import Modal from './Modal'
import Button from './Button'
import LinkButton from './LinkButton'
import Loading from './Loading'
import { Monospace } from './Text'
import { UsbDriveStatus } from '../lib/usbstick'
import USBController from './USBController'

import * as workflow from '../workflows/ExportElectionBallotPackageWorkflow'

const USBImage = styled.img`
  margin-bottom: 20px;
  height: 200px;
`

const PrimaryMessage = styled.h2`
  text-align: center;
`

const Title = styled.h1`
  margin: 0;
  margin-bottom: 20px;
  width: 100%;
  text-align: left;
`

const ExportElectionBallotPackageModalButton: React.FC = () => {
  const { electionDefinition, usbDriveStatus } = useContext(AppContext)
  const { election, electionData, electionHash } = electionDefinition!
  const electionLocaleCodes = getElectionLocales(election, DEFAULT_LOCALE)

  const [state, setState] = useState<workflow.State>(
    workflow.init(election, electionHash, electionLocaleCodes)
  )

  const [isModalOpen, setIsModalOpen] = useState<boolean>(false)

  /**
   * Execute side effects for the current state and, when ready, transition to
   * the next state.
   */
  useEffect(() => {
    ;(async () => {
      switch (state.type) {
        case 'Init': {
          setState(workflow.next)
          break
        }

        case 'ArchiveEnd': {
          await state.archive.end()
          setState(workflow.next)
        }
      }
    })()
  }, [state, election, electionData, electionHash])

  /**
   * Callback from `HandMarkedPaperBallot` to let us know the preview has been
   * rendered. Once this happens, we generate a PDF and move on to the next one
   * or finish up if that was the last one.
   */
  const onRendered = useCallback(async () => {
    if (state.type !== 'RenderBallot') {
      throw new Error(
        `unexpected state '${state.type}' found during onRendered callback`
      )
    }

    const {
      ballotStyleId,
      precinctId,
      locales,
      isLiveMode,
    } = state.currentBallotConfig
    const path = getBallotPath({
      ballotStyleId,
      election,
      electionHash,
      precinctId,
      locales,
      isLiveMode,
    })
    const data = await window.kiosk!.printToPDF()
    await state.archive.file(path, Buffer.from(data))
    setState(workflow.next)
  }, [election, electionHash, state])

  const closeModal = () => {
    setIsModalOpen(false)
    setState(workflow.init(election, electionHash, electionLocaleCodes))
  }

  const now = new Date()
  const defaultFileName = `${`${election.county.name}-${election.title}`
    .replace(/[^a-z0-9]+/gi, '-')
    .replace(/(^-|-$)+/g, '')
    .toLocaleLowerCase()}-${electionHash.slice(
    0,
    10
  )}-${now.getFullYear()}-${now.getMonth()}-${now.getDate()}_${now.getHours()}-${now.getMinutes()}-${now.getSeconds()}.zip`

  // Callback to open the file dialog.
  const openFileDialog = async () => {
    if (state.type !== 'ArchiveBegin') {
      throw new Error(
        `unexpected state '${state.type}' found during openFileDialog callback`
      )
    }
    try {
      await state.archive.begin({
        defaultPath: defaultFileName,
        filters: [{ name: 'Archive Files', extensions: ['zip'] }],
      })
      await state.archive.file('election.json', electionData)
      await state.archive.file(
        'manifest.json',
        JSON.stringify({ ballots: state.ballotConfigs }, undefined, 2)
      )

      setState(workflow.next)
    } catch (error) {
      setState(workflow.error(state, error))
    }
  }

  let mainContent = null
  let primaryButton = null
  let disableCancel = false

  switch (state.type) {
    case 'Init': {
      mainContent = <Loading as="h2" />
      break
    }

    case 'ArchiveBegin':
      switch (usbDriveStatus) {
        case UsbDriveStatus.absent:
        case UsbDriveStatus.notavailable:
        case UsbDriveStatus.recentlyEjected:
          mainContent = (
            <React.Fragment>
              <USBImage src="usb-drive.svg" alt="Insert USB Image" />
              <PrimaryMessage> No USB Drive Detected </PrimaryMessage>
              <p>
                Please insert a USB stick in order to export the ballot
                configuration.
              </p>
            </React.Fragment>
          )
          break
        case UsbDriveStatus.ejecting:
        case UsbDriveStatus.present:
          mainContent = <Loading as="h2" />
          break
        case UsbDriveStatus.mounted: {
          // TODO(caro): Update this to just write the file to the USB stick once the APIS in kiosk-browser exist.
          primaryButton = (
            <Button onPress={openFileDialog} primary>
              Export
            </Button>
          )
          mainContent = (
            <React.Fragment>
              <USBImage src="usb-drive.svg" alt="Insert USB Image" />
              <PrimaryMessage> USB Drive Detected! </PrimaryMessage>
              <p>
                Would you like to export the ballot configuration now? A zip
                archive will automatically be saved to the inserted usb drive.
                You may also {/* eslint-disable jsx-a11y/anchor-is-valid */}
                <a href="#" onClick={openFileDialog}>
                  manually select a location to save the archive to.
                </a>
                {/* eslint-enable jsx-a11y/anchor-is-valid */}
              </p>
            </React.Fragment>
          )
          break
        }
      }
      break

    case 'RenderBallot': {
      disableCancel = true
      const {
        ballotStyleId,
        precinctId,
        contestIds,
        isLiveMode,
        locales,
      } = state.currentBallotConfig
      const precinctName = getPrecinctById({ election, precinctId })!.name

      mainContent = (
        <React.Fragment>
          <PrimaryMessage>
            Generating Ballot{' '}
            {state.ballotConfigsCount - state.remainingBallotConfigs.length} of{' '}
            {state.ballotConfigsCount}…
          </PrimaryMessage>
          <ul>
            <li>
              Ballot Style: <strong>{ballotStyleId}</strong>
            </li>
            <li>
              Precinct: <strong>{precinctName}</strong>
            </li>
            <li>
              Contest count: <strong>{contestIds.length}</strong>
            </li>
            <li>
              Language format:{' '}
              <strong>{getHumanBallotLanguageFormat(locales)}</strong>
            </li>
            <li>
              Filename:{' '}
              <Monospace>{state.currentBallotConfig.filename}</Monospace>
            </li>
          </ul>
          <HandMarkedPaperBallot
            ballotStyleId={ballotStyleId}
            election={election}
            electionHash={electionHash}
            isLiveMode={isLiveMode}
            precinctId={precinctId}
            onRendered={onRendered}
            locales={locales}
          />
        </React.Fragment>
      )
      break
    }

    case 'ArchiveEnd': {
      mainContent = (
        <React.Fragment>
          <p>
            Finishing Download… Rendered{' '}
            {pluralize('ballot', state.ballotConfigsCount, true)}, closing zip
            file.
          </p>
        </React.Fragment>
      )
      break
    }

    case 'Done': {
      primaryButton = <USBController primary small={false} />
      mainContent = (
        <React.Fragment>
          <p>
            Download Complete! Exported{' '}
            {pluralize('ballot', state.ballotConfigsCount, true)}. You may now
            eject the USB device and connect it with your ballot scanning
            machine to configure it.
          </p>
        </React.Fragment>
      )
      break
    }

    case 'Failed': {
      mainContent = (
        <React.Fragment>
          <p>Download Failed! An error occurred: {state.message}.</p>
        </React.Fragment>
      )
      break
    }
  }

  return (
    <React.Fragment>
      <LinkButton small onPress={() => setIsModalOpen(true)}>
        Export Ballot Package
      </LinkButton>
      <Modal
        isOpen={isModalOpen}
        content={
          <React.Fragment>
            <Title>Export Ballot Package</Title>
            <React.Fragment>{mainContent}</React.Fragment>
          </React.Fragment>
        }
        onOverlayClick={closeModal}
        actions={
          <React.Fragment>
            <LinkButton onPress={closeModal} disabled={disableCancel}>
              Cancel
            </LinkButton>
            {primaryButton}
          </React.Fragment>
        }
      />
    </React.Fragment>
  )
}

export default ExportElectionBallotPackageModalButton
