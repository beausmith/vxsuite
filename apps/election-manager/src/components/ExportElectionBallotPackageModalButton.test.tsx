import React from 'react'
import { fireEvent, waitFor } from '@testing-library/react'
import fakeKiosk from '../../test/helpers/fakeKiosk'
import renderInAppContext from '../../test/renderInAppContext'
import ExportElectionBallotPackageModalButton from './ExportElectionBallotPackageModalButton'
import { UsbDriveStatus } from '../lib/usbstick'
import fakeFileWriter from '../../test/helpers/fakeFileWriter'

jest.mock('../components/HandMarkedPaperBallot')

beforeAll(() => {
  window.kiosk = fakeKiosk()
})

afterAll(() => {
  delete window.kiosk
})

test('Button renders properly when not clicked', async () => {
  const { container } = renderInAppContext(
    <ExportElectionBallotPackageModalButton />
  )

  expect(container).toMatchInlineSnapshot(`
    <div>
      <button
        class="sc-AxhCb dkvhmn"
        role="option"
        type="button"
      >
        Export Ballot Package
      </button>
    </div>
  `)
})

test('Modal renders insert usb screen appropriately', async () => {
  const usbStatuses = [
    UsbDriveStatus.absent,
    UsbDriveStatus.recentlyEjected,
    UsbDriveStatus.notavailable,
  ]

  for (const usbStatus of usbStatuses) {
    const { unmount, getByTestId, getByText } = renderInAppContext(
      <ExportElectionBallotPackageModalButton />,
      {
        usbDriveStatus: usbStatus,
      }
    )
    fireEvent.click(getByText('Export Ballot Package'))
    await waitFor(() => getByText('No USB Drive Detected'))

    expect(getByTestId('modal')).toMatchSnapshot()
    unmount()
  }
})

test('Modal renders export confirmation screen when usb detected', async () => {
  const { getByTestId, getByText } = renderInAppContext(
    <ExportElectionBallotPackageModalButton />,
    {
      usbDriveStatus: UsbDriveStatus.mounted,
    }
  )
  fireEvent.click(getByText('Export Ballot Package'))
  await waitFor(() => getByText('USB Drive Detected!'))

  expect(getByTestId('modal')).toMatchSnapshot()
})

test('Modal renders loading screen when usb is mounting or ejecting', async () => {
  const usbStatuses = [UsbDriveStatus.present, UsbDriveStatus.ejecting]

  for (const usbStatus of usbStatuses) {
    const { unmount, getByTestId, getByText } = renderInAppContext(
      <ExportElectionBallotPackageModalButton />,
      {
        usbDriveStatus: usbStatus,
      }
    )
    fireEvent.click(getByText('Export Ballot Package'))
    await waitFor(() => getByText('Loading'))

    expect(getByTestId('modal')).toMatchSnapshot()
    unmount()
  }
})

test('Modal renders error message appropriately', async () => {
  const { getByTestId, getByText } = renderInAppContext(
    <ExportElectionBallotPackageModalButton />,
    {
      usbDriveStatus: UsbDriveStatus.mounted,
    }
  )
  fireEvent.click(getByText('Export Ballot Package'))
  await waitFor(() => getByText('USB Drive Detected!'))

  fireEvent.click(getByText('Export'))

  await waitFor(() => getByText(/Download Failed!/))

  expect(getByTestId('modal')).toMatchSnapshot()
})

test('Modal renders renders loading message while rendering ballots appropriately', async () => {
  const mockKiosk = fakeKiosk()
  const fileWriter = fakeFileWriter()
  window.kiosk = mockKiosk
  const saveAsFunction = jest.fn().mockResolvedValue(fileWriter)
  mockKiosk.saveAs = saveAsFunction
  const { getByTestId, getByText } = renderInAppContext(
    <ExportElectionBallotPackageModalButton />,
    {
      usbDriveStatus: UsbDriveStatus.mounted,
    }
  )
  fireEvent.click(getByText('Export Ballot Package'))
  await waitFor(() => getByText('USB Drive Detected!'))

  fireEvent.click(getByText('Export'))

  await waitFor(() => getByText(/Download Complete/))
  expect(saveAsFunction).toHaveBeenCalledTimes(1)

  expect(getByTestId('modal')).toMatchSnapshot()
})
