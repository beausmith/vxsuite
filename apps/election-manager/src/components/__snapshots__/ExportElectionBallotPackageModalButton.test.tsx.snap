// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders error message appropriately 1`] = `
<div
  aria-label="Alert Modal"
  class="ReactModal__Content ReactModal__Content--after-open modal-content "
  data-testid="modal"
  role="alertdialog"
  tabindex="-1"
>
  <div
    class="sc-AxirZ djVvlL"
  >
    <h1
      class="sc-fznKkj bwmPzs"
    >
      Export Ballot Package
    </h1>
    <p>
      Download Failed! An error occurred: 
      could not begin download; no file was chosen
      .
    </p>
  </div>
  <div
    class="sc-AxjAm iIpZqc"
  >
    <button
      class="sc-AxhCb AbmWZ"
      role="option"
      type="button"
    >
      Cancel
    </button>
  </div>
</div>
`;

exports[`Modal renders export confirmation screen when usb detected 1`] = `
<div
  aria-label="Alert Modal"
  class="ReactModal__Content ReactModal__Content--after-open modal-content "
  data-testid="modal"
  role="alertdialog"
  tabindex="-1"
>
  <div
    class="sc-AxirZ djVvlL"
  >
    <h1
      class="sc-fznKkj bwmPzs"
    >
      Export Ballot Package
    </h1>
    <img
      alt="Insert USB Image"
      class="sc-fzplWN jEIXlC"
      src="usb-drive.svg"
    />
    <h2
      class="sc-fznyAO bvdqMc"
    >
       USB Drive Detected! 
    </h2>
    <p>
      Would you like to export the ballot configuration now? A zip archive will automatically be saved to the inserted usb drive. You may also 
      <a
        href="#"
      >
        manually select a location to save the archive to.
      </a>
    </p>
  </div>
  <div
    class="sc-AxjAm iIpZqc"
  >
    <button
      class="sc-AxhCb AbmWZ"
      role="option"
      type="button"
    >
      Cancel
    </button>
    <button
      class="sc-AxhCb CuUXD"
      type="button"
    >
      Export
    </button>
  </div>
</div>
`;

exports[`Modal renders insert usb screen appropriately 1`] = `
<div
  aria-label="Alert Modal"
  class="ReactModal__Content ReactModal__Content--after-open modal-content "
  data-testid="modal"
  role="alertdialog"
  tabindex="-1"
>
  <div
    class="sc-AxirZ djVvlL"
  >
    <h1
      class="sc-fznKkj bwmPzs"
    >
      Export Ballot Package
    </h1>
    <img
      alt="Insert USB Image"
      class="sc-fzplWN jEIXlC"
      src="usb-drive.svg"
    />
    <h2
      class="sc-fznyAO bvdqMc"
    >
       No USB Drive Detected 
    </h2>
    <p>
      Please insert a USB stick in order to export the ballot configuration.
    </p>
  </div>
  <div
    class="sc-AxjAm iIpZqc"
  >
    <button
      class="sc-AxhCb AbmWZ"
      role="option"
      type="button"
    >
      Cancel
    </button>
  </div>
</div>
`;

exports[`Modal renders insert usb screen appropriately 2`] = `
<div
  aria-label="Alert Modal"
  class="ReactModal__Content ReactModal__Content--after-open modal-content "
  data-testid="modal"
  role="alertdialog"
  tabindex="-1"
>
  <div
    class="sc-AxirZ djVvlL"
  >
    <h1
      class="sc-fznKkj bwmPzs"
    >
      Export Ballot Package
    </h1>
    <img
      alt="Insert USB Image"
      class="sc-fzplWN jEIXlC"
      src="usb-drive.svg"
    />
    <h2
      class="sc-fznyAO bvdqMc"
    >
       No USB Drive Detected 
    </h2>
    <p>
      Please insert a USB stick in order to export the ballot configuration.
    </p>
  </div>
  <div
    class="sc-AxjAm iIpZqc"
  >
    <button
      class="sc-AxhCb AbmWZ"
      role="option"
      type="button"
    >
      Cancel
    </button>
  </div>
</div>
`;

exports[`Modal renders insert usb screen appropriately 3`] = `
<div
  aria-label="Alert Modal"
  class="ReactModal__Content ReactModal__Content--after-open modal-content "
  data-testid="modal"
  role="alertdialog"
  tabindex="-1"
>
  <div
    class="sc-AxirZ djVvlL"
  >
    <h1
      class="sc-fznKkj bwmPzs"
    >
      Export Ballot Package
    </h1>
    <img
      alt="Insert USB Image"
      class="sc-fzplWN jEIXlC"
      src="usb-drive.svg"
    />
    <h2
      class="sc-fznyAO bvdqMc"
    >
       No USB Drive Detected 
    </h2>
    <p>
      Please insert a USB stick in order to export the ballot configuration.
    </p>
  </div>
  <div
    class="sc-AxjAm iIpZqc"
  >
    <button
      class="sc-AxhCb AbmWZ"
      role="option"
      type="button"
    >
      Cancel
    </button>
  </div>
</div>
`;

exports[`Modal renders loading screen when usb is mounting or ejecting 1`] = `
<div
  aria-label="Alert Modal"
  class="ReactModal__Content ReactModal__Content--after-open modal-content "
  data-testid="modal"
  role="alertdialog"
  tabindex="-1"
>
  <div
    class="sc-AxirZ djVvlL"
  >
    <h1
      class="sc-fznKkj bwmPzs"
    >
      Export Ballot Package
    </h1>
    <div
      class="sc-Axmtr hLmpGn"
    >
      <h2
        aria-label="Loading."
        class="sc-AxheI gcAErc"
      >
        Loading
      </h2>
    </div>
  </div>
  <div
    class="sc-AxjAm iIpZqc"
  >
    <button
      class="sc-AxhCb AbmWZ"
      role="option"
      type="button"
    >
      Cancel
    </button>
  </div>
</div>
`;

exports[`Modal renders loading screen when usb is mounting or ejecting 2`] = `
<div
  aria-label="Alert Modal"
  class="ReactModal__Content ReactModal__Content--after-open modal-content "
  data-testid="modal"
  role="alertdialog"
  tabindex="-1"
>
  <div
    class="sc-AxirZ djVvlL"
  >
    <h1
      class="sc-fznKkj bwmPzs"
    >
      Export Ballot Package
    </h1>
    <div
      class="sc-Axmtr hLmpGn"
    >
      <h2
        aria-label="Loading."
        class="sc-AxheI gcAErc"
      >
        Loading
      </h2>
    </div>
  </div>
  <div
    class="sc-AxjAm iIpZqc"
  >
    <button
      class="sc-AxhCb AbmWZ"
      role="option"
      type="button"
    >
      Cancel
    </button>
  </div>
</div>
`;

exports[`Modal renders renders loading message while rendering ballots appropriately 1`] = `
<div
  aria-label="Alert Modal"
  class="ReactModal__Content ReactModal__Content--after-open modal-content "
  data-testid="modal"
  role="alertdialog"
  tabindex="-1"
>
  <div
    class="sc-AxirZ djVvlL"
  >
    <h1
      class="sc-fznKkj bwmPzs"
    >
      Export Ballot Package
    </h1>
    <p>
      Download Complete! Exported
       
      26 ballots
      . You may now eject the USB device and connect it with your ballot scanning machine to configure it.
    </p>
  </div>
  <div
    class="sc-AxjAm iIpZqc"
  >
    <button
      class="sc-AxhCb AbmWZ"
      role="option"
      type="button"
    >
      Cancel
    </button>
    <button
      class="sc-AxhCb CuUXD"
      type="button"
    >
      Eject USB
    </button>
  </div>
</div>
`;
